// @libp2p/peer-record/dist/src/peer-record/peer-record.js
import { decodeMessage, encodeMessage, MaxLengthError, message, streamMessage } from 'protons-runtime';
import { alloc as uint8ArrayAlloc } from 'uint8arrays/alloc';
export var PeerRecord;
(function (PeerRecord) {
    let AddressInfo;
    (function (AddressInfo) {
        let _codec;
        AddressInfo.codec = () => {
            if (_codec == null) {
                _codec = message((obj, w, opts = {}) => {
                    if (opts.lengthDelimited !== false) {
                        w.fork();
                    }
                    if ((obj.multiaddr != null && obj.multiaddr.byteLength > 0)) {
                        w.uint32(10);
                        w.bytes(obj.multiaddr);
                    }
                    if (opts.lengthDelimited !== false) {
                        w.ldelim();
                    }
                }, (reader, length, opts = {}) => {
                    const obj = {
                        multiaddr: uint8ArrayAlloc(0)
                    };
                    const end = length == null ? reader.len : reader.pos + length;
                    while (reader.pos < end) {
                        const tag = reader.uint32();
                        switch (tag >>> 3) {
                            case 1: {
                                obj.multiaddr = reader.bytes();
                                break;
                            }
                            default: {
                                reader.skipType(tag & 7);
                                break;
                            }
                        }
                    }
                    return obj;
                }, function* (reader, length, prefix, opts = {}) {
                    const end = length == null ? reader.len : reader.pos + length;
                    while (reader.pos < end) {
                        const tag = reader.uint32();
                        switch (tag >>> 3) {
                            case 1: {
                                yield {
                                    field: `${prefix}.multiaddr`,
                                    value: reader.bytes()
                                };
                                break;
                            }
                            default: {
                                reader.skipType(tag & 7);
                                break;
                            }
                        }
                    }
                });
            }
            return _codec;
        };
        function encode(obj) {
            return encodeMessage(obj, AddressInfo.codec());
        }
        AddressInfo.encode = encode;
        function decode(buf, opts) {
            return decodeMessage(buf, AddressInfo.codec(), opts);
        }
        AddressInfo.decode = decode;
        function stream(buf, opts) {
            return streamMessage(buf, AddressInfo.codec(), opts);
        }
        AddressInfo.stream = stream;
    })(AddressInfo = PeerRecord.AddressInfo || (PeerRecord.AddressInfo = {}));
    let _codec;
    PeerRecord.codec = () => {
        if (_codec == null) {
            _codec = message((obj, w, opts = {}) => {
                if (opts.lengthDelimited !== false) {
                    w.fork();
                }
                if ((obj.peerId != null && obj.peerId.byteLength > 0)) {
                    w.uint32(10);
                    w.bytes(obj.peerId);
                }
                if ((obj.seq != null && obj.seq !== 0n)) {
                    w.uint32(16);
                    w.uint64(obj.seq);
                }
                if (obj.addresses != null && obj.addresses.length > 0) {
                    for (const value of obj.addresses) {
                        w.uint32(26);
                        PeerRecord.AddressInfo.codec().encode(value, w);
                    }
                }
                if (opts.lengthDelimited !== false) {
                    w.ldelim();
                }
            }, (reader, length, opts = {}) => {
                const obj = {
                    peerId: uint8ArrayAlloc(0),
                    seq: 0n,
                    addresses: []
                };
                const end = length == null ? reader.len : reader.pos + length;
                while (reader.pos < end) {
                    const tag = reader.uint32();
                    switch (tag >>> 3) {
                        case 1: {
                            obj.peerId = reader.bytes();
                            break;
                        }
                        case 2: {
                            obj.seq = reader.uint64();
                            break;
                        }
                        case 3: {
                            if (opts.limits?.addresses != null && obj.addresses.length === opts.limits.addresses) {
                                throw new MaxLengthError('Decode error - repeated field "addresses" had too many elements');
                            }
                            obj.addresses.push(PeerRecord.AddressInfo.codec().decode(reader, reader.uint32(), {
                                limits: opts.limits?.addresses$
                            }));
                            break;
                        }
                        default: {
                            reader.skipType(tag & 7);
                            break;
                        }
                    }
                }
                return obj;
            }, function* (reader, length, prefix, opts = {}) {
                const obj = {
                    addresses: 0
                };
                const end = length == null ? reader.len : reader.pos + length;
                while (reader.pos < end) {
                    const tag = reader.uint32();
                    switch (tag >>> 3) {
                        case 1: {
                            yield {
                                field: `${prefix}.peerId`,
                                value: reader.bytes()
                            };
                            break;
                        }
                        case 2: {
                            yield {
                                field: `${prefix}.seq`,
                                value: reader.uint64()
                            };
                            break;
                        }
                        case 3: {
                            if (opts.limits?.addresses != null && obj.addresses === opts.limits.addresses) {
                                throw new MaxLengthError('Streaming decode error - repeated field "addresses" had too many elements');
                            }
                            for (const evt of PeerRecord.AddressInfo.codec().stream(reader, reader.uint32(), `${prefix}.addresses[]`, {
                                limits: opts.limits?.addresses$
                            })) {
                                yield {
                                    ...evt,
                                    index: obj.addresses
                                };
                            }
                            obj.addresses++;
                            break;
                        }
                        default: {
                            reader.skipType(tag & 7);
                            break;
                        }
                    }
                }
            });
        }
        return _codec;
    };
    function encode(obj) {
        return encodeMessage(obj, PeerRecord.codec());
    }
    PeerRecord.encode = encode;
    function decode(buf, opts) {
        return decodeMessage(buf, PeerRecord.codec(), opts);
    }
    PeerRecord.decode = decode;
    function stream(buf, opts) {
        return streamMessage(buf, PeerRecord.codec(), opts);
    }
    PeerRecord.stream = stream;
})(PeerRecord || (PeerRecord = {}));
