// @libp2p/peer-store/dist/src/pb/peer.js
import { decodeMessage, encodeMessage, MaxLengthError, MaxSizeError, message, streamMessage } from 'protons-runtime';
import { alloc as uint8ArrayAlloc } from 'uint8arrays/alloc';
export var Peer;
(function (Peer) {
    let Peer$metadataEntry;
    (function (Peer$metadataEntry) {
        let _codec;
        Peer$metadataEntry.codec = () => {
            if (_codec == null) {
                _codec = message((obj, w, opts = {}) => {
                    if (opts.lengthDelimited !== false) {
                        w.fork();
                    }
                    if ((obj.key != null && obj.key !== '')) {
                        w.uint32(10);
                        w.string(obj.key);
                    }
                    if ((obj.value != null && obj.value.byteLength > 0)) {
                        w.uint32(18);
                        w.bytes(obj.value);
                    }
                    if (opts.lengthDelimited !== false) {
                        w.ldelim();
                    }
                }, (reader, length, opts = {}) => {
                    const obj = {
                        key: '',
                        value: uint8ArrayAlloc(0)
                    };
                    const end = length == null ? reader.len : reader.pos + length;
                    while (reader.pos < end) {
                        const tag = reader.uint32();
                        switch (tag >>> 3) {
                            case 1: {
                                obj.key = reader.string();
                                break;
                            }
                            case 2: {
                                obj.value = reader.bytes();
                                break;
                            }
                            default: {
                                reader.skipType(tag & 7);
                                break;
                            }
                        }
                    }
                    return obj;
                }, function* (reader, length, prefix, opts = {}) {
                    const end = length == null ? reader.len : reader.pos + length;
                    while (reader.pos < end) {
                        const tag = reader.uint32();
                        switch (tag >>> 3) {
                            case 1: {
                                yield {
                                    field: `${prefix}.key`,
                                    value: reader.string()
                                };
                                break;
                            }
                            case 2: {
                                yield {
                                    field: `${prefix}.value`,
                                    value: reader.bytes()
                                };
                                break;
                            }
                            default: {
                                reader.skipType(tag & 7);
                                break;
                            }
                        }
                    }
                });
            }
            return _codec;
        };
        function encode(obj) {
            return encodeMessage(obj, Peer$metadataEntry.codec());
        }
        Peer$metadataEntry.encode = encode;
        function decode(buf, opts) {
            return decodeMessage(buf, Peer$metadataEntry.codec(), opts);
        }
        Peer$metadataEntry.decode = decode;
        function stream(buf, opts) {
            return streamMessage(buf, Peer$metadataEntry.codec(), opts);
        }
        Peer$metadataEntry.stream = stream;
    })(Peer$metadataEntry = Peer.Peer$metadataEntry || (Peer.Peer$metadataEntry = {}));
    let Peer$tagsEntry;
    (function (Peer$tagsEntry) {
        let _codec;
        Peer$tagsEntry.codec = () => {
            if (_codec == null) {
                _codec = message((obj, w, opts = {}) => {
                    if (opts.lengthDelimited !== false) {
                        w.fork();
                    }
                    if ((obj.key != null && obj.key !== '')) {
                        w.uint32(10);
                        w.string(obj.key);
                    }
                    if (obj.value != null) {
                        w.uint32(18);
                        Tag.codec().encode(obj.value, w);
                    }
                    if (opts.lengthDelimited !== false) {
                        w.ldelim();
                    }
                }, (reader, length, opts = {}) => {
                    const obj = {
                        key: ''
                    };
                    const end = length == null ? reader.len : reader.pos + length;
                    while (reader.pos < end) {
                        const tag = reader.uint32();
                        switch (tag >>> 3) {
                            case 1: {
                                obj.key = reader.string();
                                break;
                            }
                            case 2: {
                                obj.value = Tag.codec().decode(reader, reader.uint32(), {
                                    limits: opts.limits?.value
                                });
                                break;
                            }
                            default: {
                                reader.skipType(tag & 7);
                                break;
                            }
                        }
                    }
                    return obj;
                }, function* (reader, length, prefix, opts = {}) {
                    const end = length == null ? reader.len : reader.pos + length;
                    while (reader.pos < end) {
                        const tag = reader.uint32();
                        switch (tag >>> 3) {
                            case 1: {
                                yield {
                                    field: `${prefix}.key`,
                                    value: reader.string()
                                };
                                break;
                            }
                            case 2: {
                                yield* Tag.codec().stream(reader, reader.uint32(), `${prefix}.value`, {
                                    limits: opts.limits?.value
                                });
                                break;
                            }
                            default: {
                                reader.skipType(tag & 7);
                                break;
                            }
                        }
                    }
                });
            }
            return _codec;
        };
        function encode(obj) {
            return encodeMessage(obj, Peer$tagsEntry.codec());
        }
        Peer$tagsEntry.encode = encode;
        function decode(buf, opts) {
            return decodeMessage(buf, Peer$tagsEntry.codec(), opts);
        }
        Peer$tagsEntry.decode = decode;
        function stream(buf, opts) {
            return streamMessage(buf, Peer$tagsEntry.codec(), opts);
        }
        Peer$tagsEntry.stream = stream;
    })(Peer$tagsEntry = Peer.Peer$tagsEntry || (Peer.Peer$tagsEntry = {}));
    let _codec;
    Peer.codec = () => {
        if (_codec == null) {
            _codec = message((obj, w, opts = {}) => {
                if (opts.lengthDelimited !== false) {
                    w.fork();
                }
                if (obj.addresses != null && obj.addresses.length > 0) {
                    for (const value of obj.addresses) {
                        w.uint32(10);
                        Address.codec().encode(value, w);
                    }
                }
                if (obj.protocols != null && obj.protocols.length > 0) {
                    for (const value of obj.protocols) {
                        w.uint32(18);
                        w.string(value);
                    }
                }
                if (obj.publicKey != null) {
                    w.uint32(34);
                    w.bytes(obj.publicKey);
                }
                if (obj.peerRecordEnvelope != null) {
                    w.uint32(42);
                    w.bytes(obj.peerRecordEnvelope);
                }
                if (obj.metadata != null && obj.metadata.size > 0) {
                    for (const [key, value] of obj.metadata.entries()) {
                        w.uint32(50);
                        Peer.Peer$metadataEntry.codec().encode({ key, value }, w);
                    }
                }
                if (obj.tags != null && obj.tags.size > 0) {
                    for (const [key, value] of obj.tags.entries()) {
                        w.uint32(58);
                        Peer.Peer$tagsEntry.codec().encode({ key, value }, w);
                    }
                }
                if (obj.updated != null) {
                    w.uint32(64);
                    w.uint64Number(obj.updated);
                }
                if (opts.lengthDelimited !== false) {
                    w.ldelim();
                }
            }, (reader, length, opts = {}) => {
                const obj = {
                    addresses: [],
                    protocols: [],
                    metadata: new Map(),
                    tags: new Map()
                };
                const end = length == null ? reader.len : reader.pos + length;
                while (reader.pos < end) {
                    const tag = reader.uint32();
                    switch (tag >>> 3) {
                        case 1: {
                            if (opts.limits?.addresses != null && obj.addresses.length === opts.limits.addresses) {
                                throw new MaxLengthError('Decode error - repeated field "addresses" had too many elements');
                            }
                            obj.addresses.push(Address.codec().decode(reader, reader.uint32(), {
                                limits: opts.limits?.addresses$
                            }));
                            break;
                        }
                        case 2: {
                            if (opts.limits?.protocols != null && obj.protocols.length === opts.limits.protocols) {
                                throw new MaxLengthError('Decode error - repeated field "protocols" had too many elements');
                            }
                            obj.protocols.push(reader.string());
                            break;
                        }
                        case 4: {
                            obj.publicKey = reader.bytes();
                            break;
                        }
                        case 5: {
                            obj.peerRecordEnvelope = reader.bytes();
                            break;
                        }
                        case 6: {
                            if (opts.limits?.metadata != null && obj.metadata.size === opts.limits.metadata) {
                                throw new MaxSizeError('Decode error - map field "metadata" had too many elements');
                            }
                            const entry = Peer.Peer$metadataEntry.codec().decode(reader, reader.uint32(), {
                                limits: {
                                    value: opts.limits?.metadata$value
                                }
                            });
                            obj.metadata.set(entry.key, entry.value);
                            break;
                        }
                        case 7: {
                            if (opts.limits?.tags != null && obj.tags.size === opts.limits.tags) {
                                throw new MaxSizeError('Decode error - map field "tags" had too many elements');
                            }
                            const entry = Peer.Peer$tagsEntry.codec().decode(reader, reader.uint32(), {
                                limits: {
                                    value: opts.limits?.tags$value
                                }
                            });
                            obj.tags.set(entry.key, entry.value);
                            break;
                        }
                        case 8: {
                            obj.updated = reader.uint64Number();
                            break;
                        }
                        default: {
                            reader.skipType(tag & 7);
                            break;
                        }
                    }
                }
                return obj;
            }, function* (reader, length, prefix, opts = {}) {
                const obj = {
                    addresses: 0,
                    protocols: 0,
                    metadata: 0,
                    tags: 0
                };
                const end = length == null ? reader.len : reader.pos + length;
                while (reader.pos < end) {
                    const tag = reader.uint32();
                    switch (tag >>> 3) {
                        case 1: {
                            if (opts.limits?.addresses != null && obj.addresses === opts.limits.addresses) {
                                throw new MaxLengthError('Streaming decode error - repeated field "addresses" had too many elements');
                            }
                            for (const evt of Address.codec().stream(reader, reader.uint32(), `${prefix}.addresses[]`, {
                                limits: opts.limits?.addresses$
                            })) {
                                yield {
                                    ...evt,
                                    index: obj.addresses
                                };
                            }
                            obj.addresses++;
                            break;
                        }
                        case 2: {
                            if (opts.limits?.protocols != null && obj.protocols === opts.limits.protocols) {
                                throw new MaxLengthError('Streaming decode error - repeated field "protocols" had too many elements');
                            }
                            yield {
                                field: `${prefix}.protocols[]`,
                                index: obj.protocols,
                                value: reader.string()
                            };
                            obj.protocols++;
                            break;
                        }
                        case 4: {
                            yield {
                                field: `${prefix}.publicKey`,
                                value: reader.bytes()
                            };
                            break;
                        }
                        case 5: {
                            yield {
                                field: `${prefix}.peerRecordEnvelope`,
                                value: reader.bytes()
                            };
                            break;
                        }
                        case 6: {
                            if (opts.limits?.metadata != null && obj.metadata === opts.limits.metadata) {
                                throw new MaxLengthError('Decode error - map field "metadata" had too many elements');
                            }
                            yield* Peer.Peer$metadataEntry.codec().stream(reader, reader.uint32(), `${prefix}.metadata{}`, {
                                limits: {
                                    value: opts.limits?.metadata$value
                                }
                            });
                            obj.metadata++;
                            break;
                        }
                        case 7: {
                            if (opts.limits?.tags != null && obj.tags === opts.limits.tags) {
                                throw new MaxLengthError('Decode error - map field "tags" had too many elements');
                            }
                            yield* Peer.Peer$tagsEntry.codec().stream(reader, reader.uint32(), `${prefix}.tags{}`, {
                                limits: {
                                    value: opts.limits?.tags$value
                                }
                            });
                            obj.tags++;
                            break;
                        }
                        case 8: {
                            yield {
                                field: `${prefix}.updated`,
                                value: reader.uint64Number()
                            };
                            break;
                        }
                        default: {
                            reader.skipType(tag & 7);
                            break;
                        }
                    }
                }
            });
        }
        return _codec;
    };
    function encode(obj) {
        return encodeMessage(obj, Peer.codec());
    }
    Peer.encode = encode;
    function decode(buf, opts) {
        return decodeMessage(buf, Peer.codec(), opts);
    }
    Peer.decode = decode;
    function stream(buf, opts) {
        return streamMessage(buf, Peer.codec(), opts);
    }
    Peer.stream = stream;
})(Peer || (Peer = {}));
export var Address;
(function (Address) {
    let _codec;
    Address.codec = () => {
        if (_codec == null) {
            _codec = message((obj, w, opts = {}) => {
                if (opts.lengthDelimited !== false) {
                    w.fork();
                }
                if ((obj.multiaddr != null && obj.multiaddr.byteLength > 0)) {
                    w.uint32(10);
                    w.bytes(obj.multiaddr);
                }
                if (obj.isCertified != null) {
                    w.uint32(16);
                    w.bool(obj.isCertified);
                }
                if (obj.observed != null) {
                    w.uint32(24);
                    w.uint64Number(obj.observed);
                }
                if (opts.lengthDelimited !== false) {
                    w.ldelim();
                }
            }, (reader, length, opts = {}) => {
                const obj = {
                    multiaddr: uint8ArrayAlloc(0)
                };
                const end = length == null ? reader.len : reader.pos + length;
                while (reader.pos < end) {
                    const tag = reader.uint32();
                    switch (tag >>> 3) {
                        case 1: {
                            obj.multiaddr = reader.bytes();
                            break;
                        }
                        case 2: {
                            obj.isCertified = reader.bool();
                            break;
                        }
                        case 3: {
                            obj.observed = reader.uint64Number();
                            break;
                        }
                        default: {
                            reader.skipType(tag & 7);
                            break;
                        }
                    }
                }
                return obj;
            }, function* (reader, length, prefix, opts = {}) {
                const end = length == null ? reader.len : reader.pos + length;
                while (reader.pos < end) {
                    const tag = reader.uint32();
                    switch (tag >>> 3) {
                        case 1: {
                            yield {
                                field: `${prefix}.multiaddr`,
                                value: reader.bytes()
                            };
                            break;
                        }
                        case 2: {
                            yield {
                                field: `${prefix}.isCertified`,
                                value: reader.bool()
                            };
                            break;
                        }
                        case 3: {
                            yield {
                                field: `${prefix}.observed`,
                                value: reader.uint64Number()
                            };
                            break;
                        }
                        default: {
                            reader.skipType(tag & 7);
                            break;
                        }
                    }
                }
            });
        }
        return _codec;
    };
    function encode(obj) {
        return encodeMessage(obj, Address.codec());
    }
    Address.encode = encode;
    function decode(buf, opts) {
        return decodeMessage(buf, Address.codec(), opts);
    }
    Address.decode = decode;
    function stream(buf, opts) {
        return streamMessage(buf, Address.codec(), opts);
    }
    Address.stream = stream;
})(Address || (Address = {}));
export var Tag;
(function (Tag) {
    let _codec;
    Tag.codec = () => {
        if (_codec == null) {
            _codec = message((obj, w, opts = {}) => {
                if (opts.lengthDelimited !== false) {
                    w.fork();
                }
                if ((obj.value != null && obj.value !== 0)) {
                    w.uint32(8);
                    w.uint32(obj.value);
                }
                if (obj.expiry != null) {
                    w.uint32(16);
                    w.uint64(obj.expiry);
                }
                if (opts.lengthDelimited !== false) {
                    w.ldelim();
                }
            }, (reader, length, opts = {}) => {
                const obj = {
                    value: 0
                };
                const end = length == null ? reader.len : reader.pos + length;
                while (reader.pos < end) {
                    const tag = reader.uint32();
                    switch (tag >>> 3) {
                        case 1: {
                            obj.value = reader.uint32();
                            break;
                        }
                        case 2: {
                            obj.expiry = reader.uint64();
                            break;
                        }
                        default: {
                            reader.skipType(tag & 7);
                            break;
                        }
                    }
                }
                return obj;
            }, function* (reader, length, prefix, opts = {}) {
                const end = length == null ? reader.len : reader.pos + length;
                while (reader.pos < end) {
                    const tag = reader.uint32();
                    switch (tag >>> 3) {
                        case 1: {
                            yield {
                                field: `${prefix}.value`,
                                value: reader.uint32()
                            };
                            break;
                        }
                        case 2: {
                            yield {
                                field: `${prefix}.expiry`,
                                value: reader.uint64()
                            };
                            break;
                        }
                        default: {
                            reader.skipType(tag & 7);
                            break;
                        }
                    }
                }
            });
        }
        return _codec;
    };
    function encode(obj) {
        return encodeMessage(obj, Tag.codec());
    }
    Tag.encode = encode;
    function decode(buf, opts) {
        return decodeMessage(buf, Tag.codec(), opts);
    }
    Tag.decode = decode;
    function stream(buf, opts) {
        return streamMessage(buf, Tag.codec(), opts);
    }
    Tag.stream = stream;
})(Tag || (Tag = {}));
